Rationalized user data usage

trenouf opened the pull request:
(Only the top two commits belong to this PR.)

Previously, user data usage (descriptors and push constants) worked like
this:

1. DescBuilder generated calls representing descriptor loads, push
   constant pointer
2. PatchResourceCollect noted which user data nodes are in use, and
   stored that information in ResourceUsage, an outside-IR structure.
3. PatchEntryPointMutate decided which user data nodes go in entry
   SGPRs, and stored the information in ResourceUsage
4. PatchPushConst attempted to optimize push const usage to use entry
   SGPRs instead of the spill table
5. PatchDescriptorLoad lowered the calls generated in (1) using the user
   data in SGPRs information from (3).

As part of reducing the amount of outside-IR data carried through the
passes, and also to enable a future commit that will be able to do all
this in a shader compilation with no user data layout, this commit
changes to a new scheme:

1. DescBuilder generates descriptor loads directly, in terms of calls
   representing descriptor _sets_ (and push constant pointer, and root
   descriptors as long as user data layout is available).
2. PatchEntryPointMutate directly gathers descriptor set (etc) usage,
   then decides what user data goes into entry SGPRs, then lowers the
   calls from (1) based on that. None of that usage or SGPR information
   needs to be stored outside this pass.

JaxLinAMD commented:
@trenouf , this pull request causes jenkins agents to hang, please first fix the hang on your local machine.

trenouf commented:
> @trenouf , this pull request causes jenkins agents to hang, please first fix the hang on your local machine.

OK, sorry for the trouble it caused.

linqun commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -46,26 +51,89 @@ using namespace llvm;
+
+// =====================================================================================================================
+// The entry-point mutation pass
+class PatchEntryPointMutate : public Patch {
```
why do you need remove the name space?

trenouf replied:
There is no need for out-of-line method definitions to be done inside their namespace. Also, if you define a previously-declared function (lgc::createPatchEntryPointMutate here), the LLVM coding standards doc specifically says do not do that inside the namespace, so the build catches any error earlier.

linqun commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -284,538 +352,512 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
+// * (For a GFX9+ merged shader or NGG primitive shader, the 8 system SGPRs at the start are not accounted for here.)
+// * The "user data" SGPRs, up to 32 (GFX9+ non-compute shader) or 16 (compute shader or <=GFX8). Many of the values
+//   here are pointers, but are passed as a single 32-bit register and then expanded to 64-bit in the shader code:
+//   - The "global information table", containing various descriptors such as the inter-shader rings
```
In the initial development of LLPC,  we keep user data layout close to internal compiler's layout to speed up the development.
now, I think It is time to remove unnecessary field, include per-shader table and etc.

trenouf replied:
Good idea for later commits...

linqun commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -284,538 +352,512 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
+//   - Nodes from the root user data layout, including pointers to descriptor sets. In a compute shader, these
+//     must have the same layout as the root user data layout, even if that would mean leaving gaps in register
+//     usage, and are limited to 10 SGPRs s2-s11
+//   - For a compute shader, the spill table pointer if needed, which always goes into s12. (This is possibly
```
It is a PAL limitation. it only support remap user data in graphics pipeline. I think PAL should remove this limitation in the future.

linqun commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -284,538 +352,512 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
+  // - mustSpill: The node must be loaded from the spill table; do not attempt to unspill
+
+  SmallVector<UserDataArg, 8> userDataArgs;
+  SmallVector<UserDataArg, 4> extraUserDataArgs;
```
I prefer to use specialUserDataArgs to match your comments.

linqun commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -284,538 +352,512 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
+  // - userDataArgs: global table, per-shader table and streamout table, followed by the nodes from the root user
+  //   data layout (excluding vertex buffer and streamout tables). Some of these are marked mustSpill; some of them
+  //   may need to be spilled anyway due to running out of entry SGPRs
+  // - extraUserDataArgs: special values that go at the end, such as ViewId.
```
why do we need put ViewId in the end of user datas? is it limited by PAL or just to simplify the design?
If I remember correctly, one reason is some special user data can't be put in spill table in PAL. so what will happen if we put these in the beginning of user data?

trenouf replied:
I don't know; I just copied what it did before. :-) I suspect it was to ensure it is in the same register in two shaders that are merged.

Putting at the beginning of user data should be fine in terms of not spilling. The global table and streamout table already do that.

But I propose not changing anything like that in this change.

linqun commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -284,538 +352,512 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
+  for (auto &node : m_pipelineState->getUserDataNodes()) {
+    if (node.type == ResourceNodeType::IndirectUserDataVaPtr) {
+      reserveVertexBufferTable = true;
+    } else if (node.type == ResourceNodeType::StreamOutTableVaPtr) {
```
internal compiler has removed node type StreamOutTableVaPtr, can we remove it too?

trenouf replied:
Yes -- a change I have not pushed into a PR yet stops looking in the user data layout for IndirectUserDataVaPtr (for vertex buffer) and StreamOutTableVaPtr.

linqun commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -284,538 +352,512 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
+      reserveVertexBufferTable = true;
+    } else if (node.type == ResourceNodeType::StreamOutTableVaPtr) {
+      // Only the last shader stage before fragment (ignoring copy shader) needs a stream out table.
+      if ((m_pipelineState->getShaderStageMask() &
```
Could you help to add more comment for the stream out stage check? this code is too tricky.

trenouf replied:
I changed it to use PipelineState::getLastVertexProcessingState, a method I had already created in an earlier change.

trenouf commented:
> ps: why do we need getInternalPerShaderTablePtr? if i remember correctly, we never use per pipeline table in LLPC.

You're right; nothing is using it. Perhaps I should remove it.

Can PAL cope with us not asking for it at all, and starting our user data at s1 instead of s2?

amdrexu commented on lgc/patch/SystemValues.h:
```diff
@@ -162,12 +162,9 @@ class ShaderSystemValues {
-  llvm::Value *m_spilledPushConstTablePtr = nullptr;
-  // Spilled push constant pointer
+  llvm::Instruction *m_internalGlobalTablePtr = nullptr;     // Internal global table pointer
+  llvm::Value *m_internalPerShaderTablePtr = nullptr;        // Internal per shader table pointer
```
Internal per-shader table is not used in LLPC. Maybe we can skip it now. If I remember correctly, it is still added as user_data_1 because PAL once expected us to provide it. I think you can make a separate change to have a try. If no CTS fails, it is safe to remove it and make s1 available to other uses.

trenouf replied:
Good idea for a future change. Presumably we still need to start in s2 in compute shaders.

linqun commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -284,538 +352,512 @@ bool PatchEntryPointMutate::isResourceNodeActive(const ResourceNode *node, bool
-      m_pipelineState->getPalMetadata()->setUserDataSpillUsage(node->offsetInDwords);
+    default:
+      // Any other descriptor type in the root table is accessed via the spill table.
+      mustSpill = true;
```
why do need set mustSpill for all other resource types? I think we don't have this limitations.
as far as I know, some RPM shader put the resource/sampler in root.

trenouf replied:
I think I added that limitation in an earlier change.

Because spir-v allows you to pass a pointer to an image or sampler descriptor through phi nodes and function calls, the earlier change properly separated the action of taking the pointer to the descriptor from the action of loading the descriptor.

With that separation, not spilling a root table image or sampler descriptor would require a peephole to link the two parts together. I did not implement that peephole because I thought that the only case of a root descriptor is a "dynamic" buffer descriptor.

What is an RPM shader? Some special internal shader not from the user program?

amdrexu replied:
RPM shaders are internal shaders, created in advance to execute fixed pipeline functions of graphics system. You can refer the shader sources and pipeline info in /pal/src/core/hw/gfxip/rpm/shaders from internal code base.

linqun commented on lgc/state/PalMetadata.cpp:
```diff
@@ -99,3 +136,126 @@ void PalMetadata::record(Module *module) {
+  // Set up ShaderStage -> user data register mapping.
+  m_userDataRegMapping[ShaderStageCompute] = mmCOMPUTE_USER_DATA_0;
+  m_userDataRegMapping[ShaderStageFragment] = mmSPI_SHADER_USER_DATA_PS_0;
+  if (!m_pipelineState->getNggControl()->enableNgg)
```
please group them with gfx ip. thanks!

linqun commented on lgc/state/PalMetadata.cpp:
```diff
@@ -99,3 +136,126 @@ void PalMetadata::record(Module *module) {
+         "Out of range user data index");
+
+  // Update userDataLimit if needed.
+  if (userDataValue < 0x10000 && userDataValue + dwordCount > m_userDataLimit->getUInt())
```
What's the meaning of 0x10000?

trenouf replied:
Good point. :-) I'll use a named constant and improve the comment.

linqun commented on lgc/state/PalMetadata.cpp:
```diff
@@ -99,3 +136,126 @@ void PalMetadata::record(Module *module) {
+// Set userDataLimit to maximum (the size of the root user data table, excluding vertex buffer and streamout).
+// This is called if spill is in use, or if there are root user data nodes but none of them are used (PAL does
+// not like userDataLimit being 0 if there are any root user data nodes).
+void PalMetadata::setUserDataLimitToMax() {
```
I prefer to use name setUserDataLimit. setUserDataLimitToMax is easy to causes confuse that this function will set userDataLimit to a pre-defined max value.

linqun commented on lgc/builder/DescBuilder.cpp:
```diff
@@ -39,6 +40,14 @@
 using namespace llvm;
 
+namespace {
+const unsigned DescriptorSizeResource = 8 * sizeof(unsigned);
```
I think this info should be a part of gpu property

trenouf replied:
OK, but I think I should not put the compact buffer and ycbcr sampler sizes in there as they are driver constructs and nothing to do with hardware.

trenouf commented on lgc/state/PalMetadata.cpp:
```diff
@@ -45,19 +45,40 @@ namespace {
+// consistency. That contradicts the GFX9 docs, but has the same effect.
+//
+// First the ones that only apply up to GFX8
+constexpr unsigned int mmSPI_SHADER_USER_DATA_LS_0 = 0x2D4C;
```
Using the register definitions in imported si_ci_vi_merged_offset.h and gfx9_plus_merged_offset.h is a pain, and overcomplicated given that register numbers do not change very much between architectures. Even with those imported files, there is still a mechanism to cope with what changes there are in gfx9Chip etc, and that is overcomplicated.

LLVM does not use those imported files. Given that a compiler only uses a small subset of the registers, it is better to just copy the definitions and cope with between-generation differences as and when needed. Code changes are needed for a new architecture that has register changes anyway.

So I am moving more towards the LLVM way of doing it.

Per the comment above, these particular register numbers did not change at all between gfx6-gfx10, as long as you ignore the anomaly of how the LS/HS ones are documented in gfx9.

There are separate definitions in the pipeline dumper, which cannot be compiled with LLPC/LGC code.

trenouf commented:
V4: Fixed compact buffer desc and constant indexing of root desc.
        Fixed GS spilled used data when VS merged in.
        Review comment: Put HW descriptor size consts into GPUProperty.

JaxLinAMD commented:
retest this please

JaxLinAMD commented:
llpc lit test failed.

trenouf commented:
V5: Better array of root desc and push const handling. Various fixes.
    V6: Avoid part spilling on CS fixed layout.

Reviewers might like to revisit this. :-) The top commit "Rationalized user data usage" has substantial changes to handle push const and array of root descriptors better. From the commit message:

    The unspilling algorithm has better results than the old one in a couple
    of cases:
    
    * An array of descriptors in the root table: As long as the array is not
      variably indexed, each entry is considered separately, so unused ones
      do not take an SGPR, and it might be that some entries are unspilled
      and some are spilled. (However, if the array is variably indexed, or
      this is a compute shader with fixed layout, the array is completely
      spilled.)
    * Similarly, if the push constant is a struct or array that is not
      variably indexed and is not loaded as the whole aggregate, the
      elements are considered individually. (However, if some part of the
      push constant is variably indexed, then that part is spilled, and if
      parts of the push constant are separately used on compute shader with
      fixed layout, then the whole push constant is spilled.)

trenouf commented:
Sorry, that one was completely borked, and I have probably crashed at least one of the test agents. I'll try and figure out what needs rebooting.

trenouf commented:
V8: CS fixed layout fix: don't unspill stuff after spilled desc array or
        partial-use push const

trenouf commented:
Rebased.

amdvlk-admin commented:
### Test summary for commit a8c589222d535d6f880d633465f4a6effbac2c02
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev ddcaa1009ecde219613042a5a53c6d3752bf4b9d</li>
<li>SPVGEN: dev d4817ab957f79762da83b7fc1d3f2816dbdc13fd</li>
<li>XGL: dev 278ee18285659b865ddc65f3a156c98264c8c801</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 53692d985a53a336e07907c2d4b86bf2deb66908</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 959/959 (100.0%)</li>
<li>Failed: 0/959 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

amdrexu commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -1069,6 +1506,42 @@ uint64_t PatchEntryPointMutate::pushFixedShaderArgTys(SmallVectorImpl<Type *> &a
+// For GFX9+:
+// VS -> TCS (if it exists)
+// VS -> GS (if it exists)
+// TES -> GS (if it exists)
```
Add a @param comment, describe the input stage is the original stage before merging.

amdrexu commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -1069,6 +1506,42 @@ uint64_t PatchEntryPointMutate::pushFixedShaderArgTys(SmallVectorImpl<Type *> &a
+    case ShaderStageVertex:
+      if (m_pipelineState->hasShaderStage(ShaderStageTessControl))
+        return ShaderStageTessControl;
+      LLVM_FALLTHROUGH;
```
This is very interesting. LLVM define such marker. Previously, PAL style requires a comment // Fall through for such use. But I think it is OK to use this marker. Better to write it to coding style doc

trenouf replied:
I get a warning (error with -Werror) if I do not use it.

I just checked, and I was surprised that it is not mentioned in the LLVM coding standards doc. So, yes, we should add it to our doc.

amdrexu commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -104,23 +105,121 @@ class PatchEntryPointMutate : public Patch {
+// =====================================================================================================================
+// Utility class to generate code to extend an i32 address into a 64-bit pointer, using either PC or a given
+// constant high half value.
+class PointerExtender {
```
I think the pointer is used as address, maybe rename it to AddressExtender or something similar. I just believe pointer is too general.

amdrexu commented on lgc/builder/BuilderImpl.cpp:
```diff
@@ -166,13 +166,53 @@ Instruction *BuilderImplBase::createWaterfallLoop(Instruction *nonUniformInst, A
+          base = gep->getPointerOperand();
+          continue;
+        }
+        if (auto extract = dyn_cast<ExtractValueInst>(base)) {
```
Consider changing it to extractValue. Extract/insert is OK in this context since the branch is very simple.

amdrexu commented on lgc/patch/PatchEntryPointMutate.cpp:
```diff
@@ -104,23 +105,121 @@ class PatchEntryPointMutate : public Patch {
+    // Per-descriptor-set lists of lgc.descriptor.set calls
+    SmallVector<UserDataNodeUsage, 8> descriptorSets;
+    // Minimum offset at which spill table is used.
+    unsigned spillUsage = UINT_MAX;
```
Actually, the spill threshold is set to UINT16_MAX by internal compiler :) Not an issue, but this could make the dump info consistent without any differences.

amdrexu commented on lgc/builder/DescBuilder.cpp:
```diff
@@ -243,8 +511,105 @@ Value *DescBuilder::scalarizeIfUniform(Value *value, bool isNonUniform) {
+  sqBufRsrcWord3.gfx6.dataFormat = BUF_DATA_FORMAT_32;
+  assert(sqBufRsrcWord3.u32All == 0x24FAC);
+
+  // DWORD0
```
I plan to do a change clearing some coding issues. DWORD -> Dword or dword is in the list.

trenouf replied:
I have now changed these ones.

amdvlk-admin commented:
### Test summary for commit 46ec174fd259862a27e62ebd0669bae1da113678
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev ddcaa1009ecde219613042a5a53c6d3752bf4b9d</li>
<li>SPVGEN: dev d4817ab957f79762da83b7fc1d3f2816dbdc13fd</li>
<li>XGL: dev 2ecd2e14e2ae4f2712df4a2220e2dd36265c66d1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 53692d985a53a336e07907c2d4b86bf2deb66908</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 959/959 (100.0%)</li>
<li>Failed: 0/959 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>

trenouf commented:
This passed CTS tests, but the result has not arrived back here.
